xds/src/main/java/io/grpc/xds/EdsLoadBalancer2.java

```diff
@@ -0,0 +1,419 @@
+        return false;
+      }
+      EdsLbState that = (EdsLbState) obj;
+      return Objects.equals(edsServiceName, that.edsServiceName)
+          && Objects.equals(lrsServerName, that.lrsServerName);
+    }
+
+    private final class ChildLbState extends LoadBalancer implements EndpointWatcher {
+      @Nullable
+      private final LoadStatsStore loadStatsStore;
+      private final Attributes attributes;
+      private final DropHandlingLbHelper lbHelper;
+      private List<EquivalentAddressGroup> endpointAddresses = Collections.emptyList();
+      private Map<Integer, Map<Locality, Integer>> prioritizedLocalityWeights
+          = Collections.emptyMap();
+      private PolicySelection localityPickingPolicy;
+      private PolicySelection endpointPickingPolicy;
+      @Nullable
+      private LoadBalancer lb;
+
+      private ChildLbState(Helper helper) {
+        logger.log(
+            XdsLogLevel.INFO,
+            "Start endpoint watcher on {0} with xDS client {1}", resourceName, xdsClient);
+        xdsClient.watchEndpointData(resourceName, this);
+        if (lrsServerName != null) {
+          loadStatsStore = xdsClient.addClientStats(cluster, edsServiceName);
+          xdsClient.reportClientStats();
+          attributes =
+              resolvedAddresses.getAttributes().toBuilder()
+                  .set(XdsAttributes.ATTR_CLUSTER_SERVICE_LOAD_STATS_STORE, loadStatsStore)
+                  .build();
+        } else {
+          loadStatsStore = null;
+          attributes = resolvedAddresses.getAttributes();
+        }
+        lbHelper = new DropHandlingLbHelper(helper);
+      }
+
+      @Override
+      public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+        EdsConfig config = (EdsConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+        if (lb != null) {
+          if (!config.localityPickingPolicy.equals(localityPickingPolicy)
+              || !config.endpointPickingPolicy.equals(endpointPickingPolicy)) {
+            PriorityLbConfig childConfig =
+                generatePriorityLbConfig(cluster, edsServiceName, lrsServerName,
+                    config.localityPickingPolicy, config.endpointPickingPolicy, lbRegistry,
+                    prioritizedLocalityWeights);
+            lb.handleResolvedAddresses(
+                resolvedAddresses.toBuilder()
+                    .setAddresses(endpointAddresses)
+                    .setAttributes(attributes)
+                    .setLoadBalancingPolicyConfig(childConfig)
+                    .build());
+          }
+        }
+        localityPickingPolicy = config.localityPickingPolicy;
+        endpointPickingPolicy = config.endpointPickingPolicy;
+      }
+
+      @Override
+      public void handleNameResolutionError(Status error) {
+        if (lb != null) {
+          lb.handleNameResolutionError(error);
+        } else {
+          helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
```

Comment:
This is the wrong `helper`. The right helper should be the `helper` from the constructor `ChildLbState(Helper helper)` that the `GracefulSwitchingLoadBalancer` passes to it, which could be a `pendingHelper` that does not propagate TRANSIENT_FAILURE upstream. Even forget about the above technical details, the right helper should still be the wrapped helper in the direct layer, and allow that layer to intercept update.

Same for `onResourceDoesNotExist()` and `onError()`.

Reply:
Good catch. This is a subtle one. Fixed.